src/socket/raw.rs

```diff
@@ -0,0 +1,241 @@
+use managed::Managed;
+
+use Error;
+use phy::DeviceLimits;
+use wire::{IpProtocol, Ipv4Repr, Ipv4Packet};
+use socket::{IpRepr, IpPayload, Socket};
+use storage::ring_buffer::{Resetable, RingBuffer};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PayloadType {
+    Ipv4,
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+/// A buffered raw IP packet.
+#[derive(Debug)]
+pub struct PacketBuffer<'a> {
+    size: usize,
+    payload: Managed<'a, [u8]>,
+}
+
+impl<'a> PacketBuffer<'a> {
+    /// Create new buffered packet.
+    pub fn new<T>(payload: T) -> PacketBuffer<'a>
+        where T: Into<Managed<'a, [u8]>>,
+    {
+        PacketBuffer {
+            size: 0,
+            payload: payload.into(),
+        }
+    }
+
+    fn as_ref<'b>(&'b self) -> &'b [u8] {
+        &self.payload[..self.size]
+    }
+
+    fn as_mut<'b>(&'b mut self) -> &'b mut [u8] {
+        &mut self.payload[..self.size]
+    }
+}
+
+impl<'a> Resetable for PacketBuffer<'a> {
+    fn reset(&mut self) {
+        self.size = 0;
+    }
+}
+
+/// A raw IP packet ring buffer.
+pub type SocketBuffer<'a, 'b: 'a> = RingBuffer<'a, PacketBuffer<'b>>;
+
+/// A Raw IP socket
+///
+/// A raw socket is bound to a specifict IP protocol and owns
+/// transmit and receive packet buffers.
+#[derive(Debug)]
+pub struct RawSocket<'a, 'b: 'a> {
+    payload_type: PayloadType,
+    ip_proto: IpProtocol,
```

Comment:
There is no need to have separate `payload_type` and `ip_proto` fields because the latter is strictly derived from the former.